executor/index_merge_reader.go

```diff
@@ -0,0 +1,738 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"runtime"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/ranger"
+	"github.com/pingcap/tidb/util/stringutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/zap"
+)
+
+var (
+	_ Executor = &IndexMergeReaderExecutor{}
+)
+
+// IndexMergeReaderExecutor accesses a table with multiple index/table scan.
+// There are three types of workers:
+// 1. partialTableWorker/partialIndexWorker, which are used to fetch the handles
+// 2. indexMergeProcessWorker, which is used to do the `Union` operation.
+// 3. indexMergeTableScanWorker, which is used to get the table tuples with the given handles.
+//
+// The execution flow is really like IndexLookUpReader. However, it uses multiple index scans
+// or table scans to get the handles:
+// 1. use the partialTableWorkers and partialIndexWorkers to fetch the handles (a batch per time)
+//    and send them to the indexMergeProcessWorker.
+// 2. indexMergeProcessWorker do the `Union` operation for a batch of handles it have got.
+//    For every handle in the batch:
+//	  1. check whether it has been accessed.
+//    2. if not, record it and send it to the indexMergeTableScanWorker.
+//    3. if accessed, just ignore it.
+type IndexMergeReaderExecutor struct {
+	baseExecutor
+
+	table      table.Table
+	indexes    []*model.IndexInfo
+	keepOrders []bool
+	descs      []bool
+	ranges     [][]*ranger.Range
+	dagPBs     []*tipb.DAGRequest
+	startTS    uint64
+	// handleIdx is the index of handle, which is only used for case of keeping order.
+	handleIdx    int
+	tableRequest *tipb.DAGRequest
+	// columns are only required by union scan.
+	columns           []*model.ColumnInfo
+	partialStreamings []bool
+	tableStreaming    bool
+	*dataReaderBuilder
+	// All fields above are immutable.
+
+	partialWorkerWg sync.WaitGroup
+	tblWorkerWg     sync.WaitGroup
+	processWokerWg  sync.WaitGroup
+	finished        chan struct{}
+
+	kvRanges      []kv.KeyRange
+	workerStarted bool
```

Comment:
useless `workerStarted`

Reply:
I look at the code in `IndexLookUpExecutor`. And there exists `workerStarted `. It is used in `Next()` (`executor/distsql.go:576`).

Reply:
Because it starts workers in `Next()`. So it could be removed.

Reply:
Because it starts workers in Next(). It could be removed.